foolbox/attacks/adef_attack.py

```diff
@@ -0,0 +1,405 @@
+        if `a` is a `numpy.ndarray`, must not be passed if `a` is
+        an :class:`Adversarial` instance.
+    unpack : bool
+        If true, returns the adversarial input, otherwise returns
+        the Adversarial object.
+    max_iter : int > 0
+        Maximum number of iterations (default max_iter = 100).
+    ind_of_candidates : int > 0, or array_like of int values > 0
+        The indices of labels to target in the ordering of descending
+        confidence.
+        For example:
+        - ind_of_candidates = [1, 2, 3] to target the top three labels.
+        - ind_of_candidates = 5 to to target the fifth best label.
+    max_norm : float
+        Maximum l2 norm of vector field (default max_norm = numpy.inf).
+    overshoot : float >= 1
+        Multiply the resulting vector field by this number,
+        if deformed image is still correctly classified
+        (default is overshoot = 1 for no overshooting).
+    smooth : float >= 0
+        Width of the Gaussian kernel used for smoothing.
+        (default is smooth = 0 for no smoothing).
+    targeting : bool
+        targeting = False (default) to stop as soon as model misclassifies
+        input targeting = True to stop only once a candidate label is achieved.
+
+    """
+
+    def __init__(self, model=None, criterion=Misclassification()):
+        super(ADefAttack, self).__init__(model=model, criterion=criterion)
+        self.vector_field = None
+
+    @call_decorator
+    def __call__(self, input_or_adv, ind_of_candidates=1, unpack=True,
+                 max_iter=100, max_norm=np.inf, label=None,
+                 overshoot=1.1, smooth=1.0, targeting=False):
+
+        a = input_or_adv
+        del input_or_adv
+        del label
+        del unpack
+
+        if not a.has_gradient():
+            return
+
+        # Include the correct label (index 0) in the list of targets.
+        # Remove duplicates and sort the label indices.
+        ind_of_candidates = np.unique(np.append(ind_of_candidates, 0))
+        # Remove negative entries.
+        ind_of_candidates = ind_of_candidates[ind_of_candidates >= 0]
+
+        # Number of classes to target + 1 ( >= 2).
+        # Example 1: num_classes = 10 for MNIST
+        # Example 2: If ind_of_candidates == 1,
+        # then only the second highest label is targeted
+        num_classes = ind_of_candidates.size
+
+        n = 0  # iteration number
+
+        perturbed = a.original_image.copy()  # is updated in every iteration
+
+        image_original = a.original_image.copy()  # is not updated,
+        # but kept as a copy
+
+        color_axis = a.channel_axis(batch=False)  # get color axis
+        assert color_axis in [0, 2]
+        hw = [perturbed.shape[i] for i in range(perturbed.ndim)
+              if i != color_axis]
+        h, w = hw
+
+        logits, grad, is_adv = a.predictions_and_gradient(perturbed)
+
+        logits = np.expand_dims(logits, axis=0)
```

Comment:
I don't understand why you expand the dimensions here. Wouldn't the following code be much simpler without this? Maybe there is a reason that I am overlooking right now; please explain (ideally also with a comment in the code) or simplify it if possible.

Reply:
I will simplify it. I think the only reason was that the authors of the code did it and I compared with their code line after line. I did everything in the very same way to compare. I think they actually do need this other dimension for other features. In my case, I basically left this additional dimension blank. No further reasoning here, I will remove it :)